cmd/dataobj-inspect/dump.go

```diff
@@ -0,0 +1,126 @@
+package main
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/alecthomas/kingpin/v2"
+	"github.com/fatih/color"
+	"github.com/prometheus/prometheus/model/labels"
+
+	"github.com/grafana/loki/v3/pkg/dataobj"
+	"github.com/grafana/loki/v3/pkg/dataobj/sections/logs"
+	"github.com/grafana/loki/v3/pkg/dataobj/sections/streams"
+)
+
+// dumpCommand dumps the contents of the data object.
+type dumpCommand struct {
+	files *[]string
+}
+
+func (cmd *dumpCommand) run(c *kingpin.ParseContext) error {
+	for _, f := range *cmd.files {
+		cmd.dumpFile(f)
+	}
+	return nil
+}
+
+func (cmd *dumpCommand) dumpFile(name string) {
+	f, err := os.Open(name)
+	if err != nil {
+		exitWithError(fmt.Errorf("failed to open file: %w", err))
+	}
+	defer func() { _ = f.Close() }()
+	fi, err := f.Stat()
+	if err != nil {
+		exitWithError(fmt.Errorf("failed to read fileinfo: %w", err))
+	}
+	dataObj, err := dataobj.FromReaderAt(f, fi.Size())
+	if err != nil {
+		exitWithError(fmt.Errorf("failed to read dataobj: %w", err))
+	}
+	for offset, sec := range dataObj.Sections() {
+		switch {
+		case streams.CheckSection(sec):
+			cmd.dumpStreamsSection(context.TODO(), offset, sec)
+		case logs.CheckSection(sec):
+			cmd.dumpLogsSection(context.TODO(), offset, sec)
+		default:
+			exitWithError(errors.New("unknown section"))
+		}
+	}
+}
+
+func (cmd *dumpCommand) dumpStreamsSection(ctx context.Context, offset int, sec *dataobj.Section) {
+	streamsSec, err := streams.Open(ctx, sec)
+	if err != nil {
+		exitWithError(err)
+	}
+	bold := color.New(color.Bold)
+	bold.Println("Streams section:")
+	bold.Printf("\toffset: %d\n", offset)
+
+	tmp := make([]streams.Stream, 512)
+	r := streams.NewRowReader(streamsSec)
+	for {
+		n, err := r.Read(ctx, tmp)
+		if err != nil && !errors.Is(err, io.EOF) {
+			exitWithError(err)
+		}
+		if n == 0 {
```

Comment:
This should be `n == 0 && errors.Is(err, io.EOF)` because it is possible (as far as I remember) for the reader to return 0 rows but still have more data to read
(Aside: I find myself making a mistake in these if statements so often that maybe it would be useful to have a "hasNext" method as part of the dataobj lib...)